com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs

```diff
@@ -46,11 +46,51 @@ public void Apply(TensorProxy tensorProxy, IList<int> actionIds, Dictionary<int,
         }
     }
 
+    /// <summary>
+    /// The Applier for the Discrete Action output tensor.
+    /// </summary>
+    internal class DiscreteActionOutputApplier : TensorApplier.IApplier
+    {
+        readonly ActionSpec m_ActionSpec;
+
+
+        public DiscreteActionOutputApplier(ActionSpec actionSpec, int seed, ITensorAllocator allocator)
+        {
+            m_ActionSpec = actionSpec;
+        }
+
+        public void Apply(TensorProxy tensorProxy, IList<int> actionIds, Dictionary<int, ActionBuffers> lastActions)
+        {
+            var agentIndex = 0;
+            var actionSize = tensorProxy.shape[tensorProxy.shape.Length - 1];
+            for (var i = 0; i < actionIds.Count; i++)
+            {
+                var agentId = actionIds[i];
+                if (lastActions.ContainsKey(agentId))
+                {
+                    var actionBuffer = lastActions[agentId];
+                    if (actionBuffer.IsEmpty())
+                    {
+                        actionBuffer = new ActionBuffers(m_ActionSpec);
+                        lastActions[agentId] = actionBuffer;
+                    }
+                    var discreteBuffer = actionBuffer.DiscreteActions;
+                    for (var j = 0; j < actionSize; j++)
+                    {
+                        discreteBuffer[j] = (int)tensorProxy.data[agentIndex, j];
+                    }
+                }
+                agentIndex++;
+            }
+        }
+    }
+
+
     /// <summary>
     /// The Applier for the Discrete Action output tensor. Uses multinomial to sample discrete
     /// actions from the logits contained in the tensor.
     /// </summary>
-    internal class DiscreteActionOutputApplier : TensorApplier.IApplier
+    internal class DiscreteActionWithMultiNomialOutputApplier : TensorApplier.IApplier
```

Comment:
Maybe `LegacyDiscreteActionOutputApplier`? Please update the docstring to indicate the difference between this and DiscreteActionOutputApplier

Reply:
I renamed it but the docstring is different, I added ` Uses multinomial to sample discrete
    /// actions from the logits contained in the tensor.` Is this not describing enough?

Reply:
For a new person reading the code, it wasn't clear which one is the main one - the comments were telling the "what' but not the "why". I think renaming makes it clear now.